contracts/staking/TokenStaking.sol

```diff
@@ -0,0 +1,1339 @@
+        operator.tStake -= _amount;
+        require(
+            operator.tStake >= minTStakeAmount ||
+                operator.startTStakingTimestamp + MIN_STAKE_TIME <=
+                /* solhint-disable-next-line not-rely-on-time */
+                block.timestamp,
+            "Unstaking is possible only after 24 hours"
+        );
+        emit Unstaked(_operator, _amount);
+        token.safeTransfer(operator.owner, _amount);
+    }
+
+    /// @notice Sets the legacy KEEP staking contract active stake amount cached
+    ///         in T staking contract to 0. Reverts if the amount of liquid T
+    ///         staked in T staking contract is lower than the highest
+    ///         application authorization. This function allows to unstake from
+    ///         KEEP staking contract and sill being able to operate in T
+    ///         network and earning rewards based on the liquid T staked. Can be
+    ///         called only by the delegation owner and operator.
+    function unstakeKeep(address _operator)
+        external
+        override
+        onlyOwnerOrOperator(_operator)
+    {
+        OperatorInfo storage operator = operators[_operator];
+        require(operator.owner != address(0), "Operator has no stake");
+        require(operator.keepInTStake != 0, "Nothing to unstake");
+        require(
+            getMinStaked(_operator, StakingProvider.KEEP) == 0,
+            "At least one application prevents from unstaking"
+        );
+        emit Unstaked(_operator, operator.keepInTStake);
+        operator.keepInTStake = 0;
+    }
+
+    /// @notice Reduces cached legacy NU stake amount by the provided amount.
+    ///         Reverts if there is at least one authorization higher than the
+    ///         sum of remaining legacy NU stake and liquid T stake for that
+    ///         operator or if the untaked amount is higher than the cached
+    ///         legacy stake amount. If succeeded, the legacy NU stake can be
+    ///         partially or fully undelegated on the legacy staking contract.
+    ///         This function allows to unstake from NU staking contract and
+    ///         still being able to operate in T network and earning rewards
+    ///         based on the liquid T staked. Can be called only by the
+    ///         delegation owner and operator.
+    function unstakeNu(address _operator, uint96 _amount)
+        external
+        override
+        onlyOwnerOrOperator(_operator)
+    {
+        OperatorInfo storage operator = operators[_operator];
+        require(operator.owner != address(0), "Operator has no stake");
+        require(
+            _amount > 0 &&
+                _amount + getMinStaked(_operator, StakingProvider.NU) <=
+                operator.nuInTStake,
+            "Can't unstake specified amount of tokens"
+        );
+        (, uint96 tRemainder) = tToNu(_amount);
+        operator.nuInTStake -= _amount - tRemainder;
+
+        emit Unstaked(_operator, _amount - tRemainder);
```

Comment:
I just realized there could be potential edge cases if we don't take the remainder into account before the checks. What about this?

```suggestion
        (, uint96 tRemainder) = tToNu(_amount);
        _amount -= tRemainder;
        require(
            _amount > 0 &&
                _amount + getMinStaked(_operator, StakingProvider.NU) <=
                operator.nuInTStake,
            "Can't unstake specified amount of tokens"
        );
        operator.nuInTStake -= _amount;
        emit Unstaked(_operator, _amount);
```

Reply:
good catch
✔️